GOTO Command: add issueSetGuidedModeCommand; fix related bugs

andymoe opened the pull request:
- call issueSetGuidedModeCommand before sending
  a goto command.

- Replace loop and NSThread sleepForTimeInterval:
  for retry with GCDs dispatch_after method.

- Send second commands after 500ms

- Start GCSDataManager and related Craft models to keep track of various craft state

- Factor out craft type specific mode enums

- Send quad auto and guided mode type if the craft type is not fixed wing.

- Segmented control should not function for APM:Copters

andymoe commented:
@cnatoli did one round of testing and it works as advertised. Going to do a bit more polish before asking for a review request. May fix the mode enums and add fixed wing support in a separate branch since I've already start the mode enum fixes for plane/copter elsewhere.

cnatoli commented:
@andy nice when things work out as planned!

That said, getting a fixed-wing breaking PR change past me will take more than a happy donut :-)  Particular as Guided (4) for ArduCopter maps to Acro for ArduPlane.

Let's get your other branch merged into here or master, and have a working version for fixed wing, before we merge this in.

Reading the current ArduPlane code, afaics we still want to maintain the semantics of the original code. Specifically, we do not want to send the GUIDED mode change for ArduPlane, as the mode change continues to be made implicitly when sending a guided waypoint.

As ArduPlane defaults the guided position to home (unlike ArduCopter 3.2, which defaults the guided position to the current position at the mode change), sending the mode change could result in less than desirable behaviour should the mode change transmit successfully, but the guided waypoint transmission fail.

andymoe commented:
Oops, that would be bad. I'll get it sorted for plane here sooner than
later.



On Jan 22, 2015, at 3:53 AM, Claudio Natoli <notifications@github.com>
wrote:

@andy <https://github.com/andy> nice when things work out as planned!

That said, getting a fixed-wing breaking PR change past me will take more
than a happy donut :-) Particular as Guided (4) for ArduCopter maps to Acro
for ArduPlane.

Let's get your other branch merged into here or master, and have a working
version for fixed wing, before we merge this in.

Reading the current ArduPlane code, afaics we still want to maintain the
semantics of the original code. Specifically, we do not want to send the
GUIDED mode change for ArduPlane, as the mode change continues to be made
implicitly when sending a guided waypoint.

As ArduPlane defaults the guided position to home (unlike ArduCopter 3.2,
which defaults the guided position to the current position at the mode
change), sending the mode change could result in less than desirable
behaviour should the mode change transmit successfully, but the guided
waypoint transmission fail.

—
Reply to this email directly or view it on GitHub
<https://github.com/fightingwalrus/iGCS/pull/156#issuecomment-71009696>.

andymoe commented:
@cnatoli ok, I have addressed your issues (doing my best not to crash your airplanes) and fixed related bugs around copter/plane mode mismatch. The strategy with the GCSDataManager and models is primitive at this point but it's the general direction I'd like to move toward. This branch is un-tested since verifying the basic but I should be able to test in the morning.

andymoe commented:
@cnatoli we did a test flight with APM:Copter 3.2 and have confirmed these latest changes don't introduce any regressions for copter. We will do what we can to test plane but please give the code a read and let me know what you think. The segmented control is not 100% fixed yet though it now shows auto mode when we switch to auto so there will probably be a commit or two more on this branch.

andymoe commented:
@cnatoli I just tested with the RVJet (APM 2.6) on the floor with the motor unplugged and was able to put it into guided mode so I think we are probably good.

cnatoli commented on iGCS/Comm/Interfaces/iGCS/iGCSMavLinkInterface.m:
```diff
@@ -281,18 +284,37 @@ - (void) issueRawSetWaypointCommand:(uint16_t)sequence {
+                                  coordinates.latitude, coordinates.longitude, altitude);
+
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+        mavlink_msg_mission_item_send(MAVLINK_COMM_0, msg.sysid, msg.compid, 0, MAV_FRAME_GLOBAL_RELATIVE_ALT, MAV_CMD_NAV_WAYPOINT,
```
DRY

cnatoli commented on iGCS/Comm/Interfaces/iGCS/iGCSMavLinkInterface.m:
```diff
@@ -281,18 +284,37 @@ - (void) issueRawSetWaypointCommand:(uint16_t)sequence {
+
+    mavlink_msg_set_mode_send(MAVLINK_COMM_0, msg.sysid, MAV_MODE_FLAG_CUSTOM_MODE_ENABLED, apmAutoMode);
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+         mavlink_msg_set_mode_send(MAVLINK_COMM_0, msg.sysid, MAV_MODE_FLAG_CUSTOM_MODE_ENABLED, APMCopterAuto);
```
Whoops! APMCopterAuto -> apmAutoMode.

cnatoli commented on iGCS/Comm/Interfaces/iGCS/iGCSMavLinkInterface.m:
```diff
@@ -281,18 +284,37 @@ - (void) issueRawSetWaypointCommand:(uint16_t)sequence {
+
+    mavlink_msg_set_mode_send(MAVLINK_COMM_0, msg.sysid, MAV_MODE_FLAG_CUSTOM_MODE_ENABLED, apmGuidedMode);
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+        mavlink_msg_set_mode_send(MAVLINK_COMM_0, msg.sysid, MAV_MODE_FLAG_CUSTOM_MODE_ENABLED, APMCopterGuided);
```
Ditto. APMCopterGuided -> apmGuidedMode.

cnatoli commented:
@andymoe - I'm going to branch from here, fix the couple issues in the first commit, and then try to push the GCSDataManager a bit further.

If we don't like the direction that takes, we can just cherry pick the first (fix) commit and merge this in.

cnatoli commented on iGCS/UserInterface/ViewControllers/GCSMapViewController.m:
```diff
@@ -437,14 +448,11 @@ - (void) handlePacket:(mavlink_message_t*)msg {
+
+            if (heartbeat.custom_mode == APMPlaneAuto || heartbeat.custom_mode == APMCopterAuto) {
+                idx = CONTROL_MODE_AUTO;
+            } else if (heartbeat.custom_mode == APMPlaneAuto || heartbeat.custom_mode == APMCopterAuto) {
```
Looks like another mode typo here.

Will fix on IGCS-236b branch - just marking this here in case we reject that branch.
